optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintSessionFactory.java

```diff
@@ -70,50 +72,20 @@ public DroolsConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDesc
-                (justificationList, rule) -> {
-                    DroolsConstraint<Solution_> constraint = compiledRuleToConstraintMap.get(rule);
-                    return matchJustificationsToOutput((List<Object>) justificationList,
-                            constraint.getExpectedJustificationCount(),
-                            constraint.getExpectedJustificationTypes());
-                });
-        // Determine which rules to enable based on the fact that their constraints carry weight.
-        Score<?> zeroScore = getScoreDefinition().getZeroScore();
-        Set<String> disabledConstraintIdSet = new LinkedHashSet<>(0);
-        compiledRuleToConstraintMap.forEach((compiledRule, constraint) -> {
-            Score<?> constraintWeight = constraint.extractConstraintWeight(workingSolution);
-            scoreHolder.configureConstraintWeight(compiledRule, constraintWeight);
-            if (constraintWeight.equals(zeroScore)) {
-                disabledConstraintIdSet.add(constraint.getConstraintId());
-            }
-        });
-        // Determine the KieBase to use.
-        if (disabledConstraintIdSet.isEmpty()) { // Shortcut; don't change the original KieBase.
-            currentKieBase = originalKieBase;
-            currentlyDisabledConstraintIdSet = null;
-        } else if (!disabledConstraintIdSet.equals(currentlyDisabledConstraintIdSet)) {
-            // Only rebuild the active KieBase when the set of disabled constraints changed.
-            ModelImpl model = new ModelImpl().withGlobals(originalModel.getGlobals());
-            constraintToModelRuleMap.forEach((constraintId, modelRule) -> {
-                if (disabledConstraintIdSet.contains(constraintId)) {
-                    return;
-                }
-                model.addRule(modelRule);
-            });
-            currentKieBase = KieBaseBuilder.createKieBaseFromModel(model);
-            currentlyDisabledConstraintIdSet = disabledConstraintIdSet;
-        }
-        // Create the session itself.
-        KieSession kieSession = currentKieBase.newKieSession();
-        ((RuleEventManager) kieSession).addEventListener(new OptaPlannerRuleEventListener()); // Enables undo in rules.
-        kieSession.setGlobal(DroolsScoreDirector.GLOBAL_SCORE_HOLDER_KEY, scoreHolder);
-        return new DroolsConstraintSession<>(kieSession, scoreHolder);
+        this.compiledRuleToExpectedTypesMap = compiledRuleToConstraintMap.keySet().stream()
+                .collect(Collectors.toMap(Function.identity(), rule -> {
+                    String commaSeparatedFqnList = (String) rule.getMetaData().getOrDefault(
+                            RuleAssembler.VARIABLE_TYPE_RULE_METADATA_KEY, "");
+                    return Arrays.stream(commaSeparatedFqnList.split("\\Q,\\E"))
+                            .map(className -> {
+                                try {
+                                    return Class.forName(className.trim());
```

Comment:
Where do these classNames come from?
Why are we receiving them as Strings?

Reply:
It comes from here: https://github.com/kiegroup/optaplanner/pull/820/files#diff-ebd21b451df70a2dceb49bbea389ef3eR196

It's a way of bypassing several levels of class hierarchy and the breaking of abstraction that it would cause.
I'll try once more to figure out a way to do it in a cleaner way, but I was already unsuccessful the first time around.